bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/bufferedmutator/MirroringBufferedMutator.java

```diff
@@ -114,6 +115,32 @@ public static BufferedMutator create(
   /** ExceptionListener supplied by the user. */
   protected final ExceptionListener userListener;
 
+  /** Object used as a lock to ensure ordering of scheduling flushes. */
+  private final Object flushOrderLock = new Object();
+
+  /**
+   * We have to ensure that order of asynchronously called {@link BufferedMutator#flush()} is the
+   * same as order in which callbacks for these operations were created. To enforce this property
+   * each scheduled flush will wait for previously scheduled flush to finish before performing its
+   * operation. We are storing futures of last scheduled flush operation in this field.
+   *
+   * <p>Access to this field should be synchronized by {@code synchronized(flushOrderLock)}
+   *
+   * <p>We have to ensure the ordering to prevent the following scenario:
+   *
+   * <ol>
+   *   <li>main thread: scheduleFlush with callback using dataToFlush = [1,2,3] (flush1)
+   *   <li>main thread: scheduleFlush with callback using dataToFlush = [] (flush2).
+   *   <li>worker thread 1: call flush2, it blocks.
+   *   <li>worker thread 2: call flush1, nothing more to flush (there's no guarantee that this flush
+   *       would wait for flush2 to finish), callback with dataToFlush = [1,2,3] is called before
+   *       corresponding mutations were flushed.
+   * </ol>
+   *
+   * Ensuring the order of flushes forces to be run after flush1 is finished.
+   */
+  private FlushFutures lastFlushFutures = createCompletedFlushFutures();
```

Comment:
How about we pack `createCompletedFlushFutures` and `flushOrderLock` into a class, e.g. `FlushSerializer`? It would have to take callbacks, but you would contain the synchronization primitives to a single place.

Reply:
Done.